frontends/p4/defaultValues.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2023 VMWare, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "defaultValues.h"
+
+namespace P4 {
+
+// static
+const IR::Expression *DoDefaultValues::defaultValue(Util::SourceInfo srcInfo,
+                                                    const IR::Type *type) {
+    if (auto tb = type->to<IR::Type_Bits>()) {
+        return new IR::Constant(srcInfo, tb, 0);
+    } else if (type->is<IR::Type_InfInt>()) {
+        return new IR::Constant(srcInfo, 0);
+    } else if (type->is<IR::Type_Boolean>()) {
+        return new IR::BoolLiteral(srcInfo, false);
+    } else if (auto te = type->to<IR::Type_Enum>()) {
+        return new IR::Member(srcInfo, new IR::TypeNameExpression(te->name),
+                              te->members.at(0)->getName());
+    } else if (auto te = type->to<IR::Type_SerEnum>()) {
+        return new IR::Cast(srcInfo, type->getP4Type(), new IR::Constant(srcInfo, te->type, 0));
+    } else if (auto te = type->to<IR::Type_Error>()) {
+        return new IR::Member(srcInfo, new IR::TypeNameExpression(te->name), "NoError");
+    } else if (type->is<IR::Type_String>()) {
+        return new IR::StringLiteral(srcInfo, cstring(""));
+    } else if (type->is<IR::Type_Varbits>()) {
+        ::error(ErrorType::ERR_UNSUPPORTED, "%1% default values for varbit types", srcInfo);
+        return nullptr;
+    } else if (auto ht = type->to<IR::Type_Header>()) {
+        return new IR::InvalidHeader(ht->getP4Type());
+    } else if (auto hu = type->to<IR::Type_HeaderUnion>()) {
+        return new IR::InvalidHeaderUnion(hu->getP4Type());
+    } else if (auto st = type->to<IR::Type_StructLike>()) {
+        auto vec = new IR::IndexedVector<IR::NamedExpression>();
+        for (auto field : st->fields) {
+            auto value = defaultValue(srcInfo, field->type);
```

Comment:
What happens here if `field->type` is a Type_Name? Is this resolved beforehand somehow? If so, how?

Reply:
It really depends whether that is the canonical type (from the typeMap) or not.
If it's the canonical type there should never be a Type_Name.
But it does make the point that the previous version of this function is not equivalent to the current one.
This will crash rather than be wrong, so let's leave it like this for now.

Reply:
How do you retrieve the canonical type? `getTypeType` or `getType`? 
I would like to use this function so I have to make sure the input type is canonical.

Reply:
When you get a type from the typeMap it is canonical, and so are all it's subfields.
if you get a type from the program it is probably not canonical.
A canonical type can be converted to a program type with getP4Type().